src/TTImp/ProcessData.idr

```diff
@@ -462,17 +463,25 @@ processData {vars} eopts nest env fc vis mbtot (MkImpData dfc n_in mty_raw opts
          let metas = maybe empty fst mmetasfullty
          let mfullty = map snd mmetasfullty
 
-         -- If n exists, check it's the same type as we have here, and is
-         -- a data constructor.
+         -- If n exists, check it's the same type as we have here, is
+         -- a type constructor, and has either the same visibility or we don't define one.
          -- When looking up, note the data types which were undefined at the
          -- point of declaration.
          ndefm <- lookupCtxtExact n (gamma defs)
          (mw, vis, fullty) <- the (Core (List Name, Visibility, ClosedTerm)) $ case ndefm of
                   Nothing => case mfullty of
                     Nothing => throw (GenericMsg fc "Missing telescope for data definition \{show n_in}")
-                    Just fullty => pure ([], vis, fullty)
-                  Just ndef =>
-                    let vis = max (visibility ndef) vis in
+                    Just fullty => pure ([], collapseDefault def_vis, fullty)
+                  Just ndef => do
+                    vis <- the (Core Visibility) $ case def_vis of
+                      Default => pure ndef.visibility
+                      Value newVis =>
+                        if newVis == ndef.visibility
+                        then pure newVis
+                        -- TODO : In a later release, replace this with an error.
+                        else do recordWarning (IncompatibleVisibility fc ndef.visibility newVis n)
+                                pure (max ndef.visibility newVis)
```

Comment:
I don't think that's the correct logic here and that's why you had to add some `public export`s to
some of the `libs/papers` modules.

What you want is something like the following that attempts to merge two withdefaults and:

* Fails if they're incompatible values
* Happily overwrites any `Default

```
mergeValues : Eq a => {def : a} -> WithDefault a def -> WithDefault a def ->
			Either (a, a) (WithDefault a def)
mergeValues (Value x) (Value y) = if x == y then Left (x, y) else Right (Value x)
mergeValues (Value x) Default = Right (Value x)
mergeValues Default w = Right w
```

Once you have that you can check whether `def_vis` and the `ndef`'s withdefault are compatible.

Reply:
So you mean to allow
```idris
public export
data Foo : Type
data Foo : Type where
```
and
```idris
data Foo : Type
public export
data Foo : Type where
```
to both mean the same? In that case, I can add that in tomorrow, will just result in a bigger overall change once I modify `GlobalDef` as well.

Reply:
Yes I think so. It's indeed a bit more invasive but gives users more freedom.

Reply:
Added! I left `IClaim`s alone/not carrying the default-info since we don't have a "IClaimLater/IClaimImp" distinction like with records and data declarations.